gui/test/app/tinystan/model.test.ts

```diff
@@ -0,0 +1,427 @@
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      mockedModule.stringToUTF8.mockClear();
+
+      model.sample({ data: dataStr });
+      expect(mockedModule.stringToUTF8).toHaveBeenNthCalledWith(
+        1,
+        dataStr,
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      expect(mockedModule._tinystan_sample).toHaveBeenCalledTimes(2);
+    });
+
+    test("multiple inits are passed correctly", async () => {
+      const { mockedModule, model } = await getMockedModel({});
+
+      const inits = { hi: 1 };
+      const initsStr = JSON.stringify(inits);
+
+      model.sample({ inits });
+
+      expect(mockedModule.stringToUTF8).toHaveBeenCalledWith(
+        initsStr,
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      mockedModule.stringToUTF8.mockClear();
+
+      const inits2 = [inits, inits];
+      const inits2Str = initsStr + "%" + initsStr;
+
+      model.sample({ inits: inits2, num_chains: 2 });
+
+      expect(mockedModule.stringToUTF8).toHaveBeenCalledWith(
+        inits2Str,
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      expect(mockedModule._tinystan_sample).toHaveBeenCalledTimes(2);
+    });
+
+    test("output is correct shape and layout", async () => {
+      const { model } = await getMockedModel({
+        paramNames: "a,b",
+      });
+
+      const num_chains = 3;
+      const num_samples = 5;
+      const num_warmup = 7;
+      const N_PARAMS = HMC_SAMPLER_VARIABLES.length + 2;
+
+      // Two cases:
+      // 1. save_warmup = false
+      //    - draws is num_params x (num_chains * num_samples)
+      // 2. save_warmup = true
+      //    - draws is num_params x (num_chains * (num_samples + num_warmup))
+      {
+        const { draws, paramNames } = model.sample({
+          num_chains,
+          num_samples,
+          num_warmup,
+          save_warmup: false,
+          adapt: false,
+        });
+
+        expect(paramNames).toContain("a");
+        expect(paramNames).toContain("b");
+        expect(draws.length).toEqual(paramNames.length);
+
+        expect(draws.length).toEqual(N_PARAMS);
+        draws.forEach((d) =>
+          expect(d.length).toEqual(num_chains * num_samples),
+        );
+
+        expectColumnMajor(draws, N_PARAMS, num_chains * num_samples);
+      }
+
+      {
+        const { draws, paramNames } = model.sample({
+          num_chains,
+          num_samples,
+          num_warmup,
+          save_warmup: true,
+        });
+
+        expect(paramNames).toContain("a");
+        expect(paramNames).toContain("b");
+        expect(draws.length).toEqual(paramNames.length);
+
+        expect(draws.length).toEqual(N_PARAMS);
+        draws.forEach((d) =>
+          expect(d.length).toEqual(num_chains * (num_samples + num_warmup)),
+        );
+
+        expectColumnMajor(
+          draws,
+          N_PARAMS,
+          num_chains * (num_samples + num_warmup),
+        );
+      }
```

Comment:
It's not terribly important, but you could probably do this as a loop over `[true, false]` for `save_warmup`